### ISSUE TITLE
XLNet: Generalized Autoregressive Pretraining for Language Understanding

### ISSUE THREAD

**Issue by seopbo** [ModuNLP/weekly-meeting#52]
## 어떤 내용의 레퍼런스인가요? 👋

간략하게 이 글이 어떤 내용인지 작성해 주세요! (짧게 1-2줄 이어도 좋아요!)

## 이 글을 읽어서 무엇을 배울 수 있는지 알려주세요! 🤔

이 글을 제대로 읽었을 때 어떤 지식을 얻을 수 있을까요?

## 같이 읽어보면 좋을 만한 글이나 이슈가 있을까요?

만약에 있다면 자유롭게 작성해 주세요!

## 레퍼런스의 URL을 알려주세요! 🔗
- https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_xlnet.py
- https://arxiv.org/abs/1906.08237

**Comment by eagle705** [comment]
## 어떤 내용의 레퍼런스인가요? 👋

BERT 이후 SOTA를 기록했던 PLM인 XLNET에 대해 알아보자
- 저자: Zhilin Yang, ... (CMU + Google AI Brain Team)
- 인용수: 1,000 이상
- Conference: NeurIPS 2019

## 이 글을 읽어서 무엇을 배울 수 있는지 알려주세요! 🤔

### XLNET 이전의 BERT 모델에서의 문제
- BERT는 ```[MASK]``` 라는 noise를 입력으로 넣고, output에 원래의 값을 복원하는 denoising autoencoder (AE) 방식임
- Pretraining 학습에는 항상 Noise가 들어가지만, Finetuning에서는 이러한 노이즈가 들어가지 않기 때문에 Pretraining과 Finetuning시 입력의 불일치 발생
- BERT의 경우 ```[MASK]``` 토큰 자리에 있던 정답 토큰과 input sequence안에 있는 토큰들이 함께 등장할 확률이 독립이 아님에도 불구하고 독립으로 가정함
  - 각 토큰간의 관계가 독립이 아니므로, 이것을 모델링할 수 있는 flow(?)가 필요함

### XLNet은 어떻게 동작하나?
- 기존 LM은 Autogressive (AR) 방식임
- 양방향 문맥파악이 어려움 (ELMo도 각각 방향에 대해 독립학습이라..)
- token index를 permutation하고 AR방식을 적용함으로써 AE와 비슷한 효과를 누릴 수 있게 함
- [기창님 블로그 설명 (id가 3인 토큰만 맞춘다고 생각하고 보면됨)](https://ratsgo.github.io/natural%20language%20processing/2019/09/11/xlnet/)
- 실제로는 index를 permutation하기보다 attention mask로 permutation 효과를 적용
- Query token position의 token도 보여줌으로써 Context를 보는 Attention Stream과 기존 MLM처럼 Query token position의 token은 숨기되, token position 정보를 알려줘서 target token을 맞추는 Attention Stream 을 사용함으로써 문제를 해결하고려함

### XLNet의 Contribution
- ```[MASK]``` 토큰을 사용하지 않고 모델링함으로써, pretraining과 finetuning간의 불일치 해소
- Two-Stream Self-Attention를 통한 Target Aware Representation을 제안함으로써, 각 단어들 끼리는 서로간에 dependency를 고려하면서도 target 단어를 맞추는 LM을 만듦
- Tranformer-XL 구조를 사용함으로써 긴 문서에 대해서 대응할 수 있게됨 (Memory 구조 도입)


#### [XLNet 구조](https://github.com/huggingface/transformers/blob/31da35cc8939e19a292cb7f15cad5c9a1ddf7f23/src/transformers/modeling_xlnet.py#L910)
- Backbone: Transformer
- Embedding: 
  - [token embedding](https://github.com/huggingface/transformers/blob/31da35cc8939e19a292cb7f15cad5c9a1ddf7f23/src/transformers/modeling_xlnet.py#L1145)
  - [relative pos embedding](https://github.com/huggingface/transformers/blob/31da35cc8939e19a292cb7f15cad5c9a1ddf7f23/src/transformers/modeling_xlnet.py#L1172)
  - [segment embedding](https://github.com/huggingface/transformers/blob/31da35cc8939e19a292cb7f15cad5c9a1ddf7f23/src/transformers/modeling_xlnet.py#L1157)
- Memory:
  - [이전 state + 현재 word embdding](https://github.com/huggingface/transformers/blob/31da35cc8939e19a292cb7f15cad5c9a1ddf7f23/src/transformers/modeling_xlnet.py#L1201)
- [layer_module:](https://github.com/huggingface/transformers/blob/31da35cc8939e19a292cb7f15cad5c9a1ddf7f23/src/transformers/modeling_xlnet.py#L492)
  - [relative attention:](https://github.com/huggingface/transformers/blob/31da35cc8939e19a292cb7f15cad5c9a1ddf7f23/src/transformers/modeling_xlnet.py#L210)
    - input: ``` cat = torch.cat([mems, h], dim=0)``` 
    - einsum을 통해 QKV 연산
    - output: output_h, output_g (Context Representation, Query Representation)
  - [feed forward:](https://github.com/huggingface/transformers/blob/31da35cc8939e19a292cb7f15cad5c9a1ddf7f23/src/transformers/modeling_xlnet.py#L469)
    - linear - act - dropout - linear - dropout - layer_norm
  - dropout

### permutation
- https://github.com/zihangdai/xlnet/blob/master/data_utils.py#L597

## 같이 읽어보면 좋을 만한 글이나 이슈가 있을까요?

- Context Stream을 Loss에 어떻게 반영하는지가 잘 이해가 안됨..

## 레퍼런스의 URL을 알려주세요! 🔗
- 기창님 블로그: https://ratsgo.github.io/natural%20language%20processing/2019/09/11/xlnet/
- 핑퐁 블로그: https://blog.pingpong.us/xlnet-review/
